### DISCUSSION TITLE
Filecoin 扇区生命周期及状态管理

### DISCUSSION THREAD

**Discussion by maxvint** [filecoin-project/community-china#14] [新手上路]
# 扇区生命周期及状态管理

扇区是Filecoin网络中的数据存储单元，目前主网的扇区大小有32GiB和64GiB。

## 1. 扇区生命周期详解
Filecoin网络的扇区，需要通过一系列的计算过程，最终得到扇区内数据的证明结果，并存储到区块链上。

扇区的主要计算过程包括：PreCommit1(PC1)、PreCommit2(PC2)、Commit2(C2)三个过程，目前在AMD EPYC 7542 CPU 和 2080Ti显卡上，基于官方代码，计算PC1的过程大约需要3个小时，PC2需要25分钟，C2需要40分钟。
基于社区及优化后的代码，PC2为10分钟左右，C2在15分钟左右，社区优化版的链接如下：
- [P2 优化版](https://github.com/filguard/rust-fil-proofs)
- [C2 优化版](https://github.com/jackoelv/bellperson)

### 1.1 随机数扇区生命周期
随机数扇区即通过`lotus-miner sectors pledge`生成的扇区，扇区中存储的都是没有实际价值的随机数，目前仅仅是为了承诺有效算力。可以通过操作，将随机数扇区声明为有效数据存储的扇区。

### 1.2 订单扇区生命周期
订单扇区是矿工通过Filecoin的订单撮合市场，存储了客户的真实数据以后生成的扇区。

## 2. 扇区状态管理
扇区主要状态包括：PreCommit1、PreCommit2、Committing、FinalizeSector，状态变化如下图所示：

<img src="https://user-images.githubusercontent.com/1715211/110258389-570e7f80-7fdd-11eb-9654-2c0e6dfa9ba2.png" width="680" />

### 2.1 更新扇区状态
```sh
$ lotus-miner sectors update-state --really-do-it <sectorId> <newState>
```
针对`CommitFailed`状态的扇区，可通过以下命令，将扇区状态更改为`Committing`状态。
```sh
$ lotus-miner sectors update-state --really-do-it <sectorId> Committing
```

### 2.2 删除本地的扇区
```sh
$ lotus-miner sectors remove --really-do-it <sectorId>
```
针对`PreCommitFailed`和`SealPreCommit1Failed`状态的扇区，因为还没有质押，可通过上面的命令直接删除。

**关于如何正确删除扇区，请参考：**
- [如何正确删除扇区](https://github.com/filecoin-project/community-china/blob/master/documents/tutorial/How_to_delete_sector_correctly/How_to_delete_sector_correctly.md)
- [顽固扇区如何删除](https://github.com/filguard/lotus-ops/blob/master/documents/questions.md#1-%E9%A1%BD%E5%9B%BA%E6%89%87%E5%8C%BA%E5%A6%82%E4%BD%95%E5%88%A0%E9%99%A4)

### 2.3 删除链上的扇区
对于已经上链的扇区，如果数据丢失或者恢复失败(RecoveryTimeout)，需要用下面的命令在链上销毁掉。
```sh
lotus-shed sectors terminate --really-do-it <sectorId-1> <sectorId-2>...
```
一次可以批量销毁多个扇区，**注意，链上销毁扇区会有惩罚，每销毁一个扇区大概要惩罚0.1个币**。

**Comment by 0xhhhhh** [comment]
文中有提到，基于社区及优化后的代码，请问哪里可以下载呢？

**Reply by DancGo** [reply]
+同问
哪里可以下载到“基于社区及优化后的代码”？

**Reply by jackyqj** [reply]
+同问

**Reply by KaixuanZhao** [reply]
IPFS主网最大雅典娜矿池，提供一站式采购托管运营，fil存量数据迁移出海等服务，全网算力领先1400PB，需要B端1PB以上，C端50T，100T，200T（可扩容）和FIL存量数据迁移出海服务方案的滴滴我，欢迎加入社群Wechet：kkmmjjnn00

**Comment by zhanglei-githb** [comment]
+同问
哪里可以下载到“基于社区及优化后的代码”？

**Comment by zhangrui231** [comment]
+同问
哪里可以下载到“基于社区及优化后的代码”？

**Comment by arechan** [comment]
+同问
哪里可以下载到“基于社区及优化后的代码”？

**Comment by galaxy0x** [comment]
我有个扇区卡在GetTicket 不管是lotus-miner sectors update-state --really-do-it还是lotus-miner sectors remove --really-do-it都是报ERROR: normal shutdown of state machine。这是怎么回事呀？

**Reply by linwujia** [reply]
你这个属于扇区对应的状态机已经关闭了，需要重启miner才能继续

**Reply by linwujia** [reply]
你这个属于扇区对应的状态机已经关闭了，需要重启miner才能继续

**Comment by liufengpkw** [comment]
同问优化后的代码哪里下载，有联系方式没给一个

**Comment by KaixuanZhao** [comment]
RRMine，雅典娜矿池，全网算力领先1000PB，需要B端C端方案的滴滴我，交流学习，欢迎加入社群Wechet：kkmmjjnn00

**Comment by yhhwpp** [comment]
基于社区及优化后的代码，不说重点。

**Comment by maxvint** [comment]
@yhhwpp @liufengpkw @arechan @zhangrui231 社区的优化版本：
[P2 优化版](https://github.com/filguard/rust-fil-proofs)
[C2 优化版](https://github.com/jackoelv/bellperson)

**Comment by kevinlv1024** [comment]
基于社区优化后的代码 [P2](https://github.com/filguard/rust-fil-proofs)   404 Page not Found  ，有人还有这个源码吗？

**Comment by zhaobo443** [comment]
FIL联合封装，用户提供质押币，公司提供设备和运维，收益60％,公司扣出20％服务费和20%设备费，合同期1080天，540天质押币退还，t3天封装。可实地考察。V:zhaobo9308

**Comment by guanxinlian** [comment]
FIL联合封装，用户提供质押币，公司提供设备和运维、自建机房，收益5:5，540天质押币退还。可实地考察（嘉兴）。V:  guan71439
